### DISCUSSION TITLE
Comparison to Dynamoose

### DISCUSSION THREAD

**Discussion by bestickley** [tywalch/electrodb#212] [General]
Hi, how does electrodb compare to [dynamoose](https://dynamoosejs.com/getting_started/Introduction)?

**Comment by ikushlianski** [comment]
To me the key differences were:
1. Single-Table design support
2. Ability to create resources on AWS automatically

**1. Single-Table design support**
I think Dynamoose doesn't have any support for Single-Table Design. They mention Single-Table Design only once in their docs as far as I noticed, on this page https://dynamoosejs.com/guide/Model.

With Dynamoose, you have to manually model your rows, including separations of entities with `#` and values with `_`. When you query your table, all these separators will get returned to you, so you will have to do string manipulations yourself when mapping from DB layer to your domain layer. ElectroDB returns to you only pure data without these ugly separators, although they are present in Dynamo raw data.

ElectroDB is focused strongly on Single-Table Design and does a lot of work under the hood for it, so we can build our access patterns as indexes and give them a recognizable name. It also auto-builds your PK and SK to improve experience of fetching an item with its related items in one call (it's called `collections`, which simulate relational DB joins). All of this is absent from Dynamoose. You can certainly do all that manually with Dynamoose, because in the end it is just plain DynamoDB under the hood, but the learning curve for the dev team will be higher.

**Ability to create resources on AWS automatically**
Dynamoose seems to be able to create tables automatically for you if they don't exist. If you are using infrastructure as code, you probably won't need this. ElectroDB assumes you already have your tables and indexes in place and doesn't create anything for you.

There is certainly a lot more about the differences, but these were the key ones to me. We were looking for something with Single-Table design support and ElectroDB seems a great fit for now.

**Reply by bestickley** [reply]
Great explanation, thanks @ikushlianski!
